### PR TITLE
copy is removed from 1.24 

### PR DESCRIPTION
**Pull request by velo**
Fix for https://github.com/spotify/docker-client/issues/465

### CODE REVIEW

**Comment by KostyaSha** [comment]
How would apiVersion check work with swarm?

**Comment by velo** [comment]
@KostyaSha I guess the way it is already....

I made no changes to it

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 1]
```diff
@@ -729,6 +729,21 @@ public InputStream exportContainer(String containerId)
   @Override
   public InputStream copyContainer(String containerId, String path)
       throws DockerException, InterruptedException {
+
+    final String apiVersion = version().apiVersion();
+    final int versionComparison = compareVersion(apiVersion, "1.20");
+
+    // Version below 1.20
+    if (versionComparison < 0) {
+      return doCopyContainer(containerId, path);
```
Comment (johnflavin) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: This doesn't seem like the right design to me. I would not expect the `copyContainer` method to do different things depending on the version of docker you’re connected to. I would think there would be two different methods for the two different API endpoints: `copyContainer` that wraps `/containers/{id}/copy` and `archiveContainer` that wraps `/containers/{id}/archive`. `DockerClient.copyContainer` should be marked `@Deprecated`.

**Comment by mattnworb** [comment]
I agree with @johnflavin. Also #465 mentions API version 1.24 but this PR checks 1.20 - why is that?

**Comment by mattnworb** [comment]
@KostyaSha I am not very familiar with Swarm, but how does that related to checking the version of the Docker Remote API used to communicate with the docker daemon?

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java (hunk 1 above)
Comment (velo) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: ok, I can do that, thanks

**Comment by velo** [comment]
This option is available since 1.20 why is why.... but mandatory only on 1.24m anyway, I will apply the asked change witch will pretty much remove this

**Comment by velo** [comment]
@johnflavin @mattnworb done!

File: src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.docker.client.exceptions;
+
+public class UnsupportedApiVersionException extends NotFoundException {
```
Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java]: why does this extend NotFoundException?

Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java]: why does this extend NotFoundException? That seems confusing and unnecessary.

Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java]: Why does this extend NotFoundException? That seems unnecessary.

**Comment by mattnworb** [comment]
Thanks @velo for making the requested updates. My only remaining concern is that the UnsupportApiVersionException extends NotFoundException - why is that? It doesn't seem as if the two are related.

I'd be happy to merge if that issue was taken care of.

File: src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java (hunk 2 above)
Comment (velo) [inline src/main/java/com/spotify/docker/client/exceptions/UnsupportedApiVersionException.java]: My bad, fixed

**Comment by velo** [comment]
@mattnworb done

**Comment by mattnworb** [comment]
@velo thanks!

**Comment by quickwind** [comment]
@mattnworb When will this be released? Seems no new version since Jun 8....

**Comment by F-Scippacercola** [comment]
Hi,

I was looking to the corrections made in the API due to the deprecation of the copy endpoint.

Instead of just making the interface of copyCointainer fail, when docker version is >= 1.24, I would suggest to forward the invocation of copy to the new archive endpoint, but preserving the interface of the original call, in order to keep the library backward-compatible. 

Moreover, it seems that the new archive endpoint does not return all the exceptions of the new API, such as "NotFoundException", that should be handled by the library.

**Comment by velo** [comment]
Sure, go for it

**Comment by johnflavin** [comment]
I don’t agree that `copy` should silently redirect to `archive` depending on the version of docker you are connected to. In fact, that was the original design of this PR and I argued against it [here](https://github.com/spotify/docker-client/pull/466#discussion_r69149083). My opinion is that the methods in `docker-client` should reflect the docker REST API endpoints they connect to. So when they deprecate `copy` in favor of `archive`, we do the same.

**Pull request merged**